### ISSUE TITLE
Dynamically rotate labels based on cell and label widths

### ISSUE THREAD

**Issue by bye7** [amcharts/amcharts5#768]
**Problem**
I am unable to use `xAxis.events.on("sizechanged", function () { })` on amCharts 5. I believe it may be only available on [v4](https://www.amcharts.com/docs/v4/reference/sprite/#sizechanged_event) as that type of event is possibly not being listened for. How could I access to a similar event as I want to dynamically rotate labels on the xAxis depending on the cellWidth and label width.

**Environment**
amCharts version: 5

**Additional context**
I am trying to dynamically rotate labels on the xAxis depending on the cellWidth and label width value.

**Question**
How can I access to a similar event to **xAxis.events.on("sizechanged", function () { })** in amCharts 5?

**Comment by martynasma** [comment]
There's an easier way to track changes in size:

```JavaScript
xAxis.onPrivate("width", function(width, target) {   });
xAxis.onPrivate("height", function(height, target) {   });
```

**Comment by bye7** [comment]
I'm not sure if you can get the exact same result from the [amCharts 4](https://codepen.io/team/amcharts/embed/GRRoamd) example by accessing the width private setting. It would not work if you were to zoom-in or add data to the chart, because that changes the grid granularity (were labels can appear close together depending on `minGridDistance` property value) and not width of the chart. We might be able to use
```JS
xRenderer.grid.template.events.on("boundschanged", function (target) { });
```
then figure out the grid line distance which we could use as a gauge for label rotation.

It's a lot confusing when an event is removed that was working in the previous version. It would be nice if a feature like this could be directly implemented as a property setting like in Chart.js that automatically rotates labels:  

https://user-images.githubusercontent.com/82799248/215013350-ab05080c-aa4d-4e35-b24b-7fcf90fa6b64.mp4

https://user-images.githubusercontent.com/82799248/215013327-efae4af3-6f5b-4972-94f0-d32ed6cc257b.mp4

I've now moved on to "XY chart (duration axis)" and realized that dynamic data item grouping is not working for `DurationAxis`, although it worked for `DateAxis`. It was a great feature that greatly improved performance.

I'm also unable to override `durationFormats` setting. I believe `root.numberFormatter` does not have `durationFormats` property, so the following code from the [docs](https://www.amcharts.com/docs/v5/concepts/formatters/formatting-durations/#Setting_default_format) returns `Cannot read properties of undefined` error:
```JS
root.numberFormatter.get("durationFormats")["hour"] = {
  hour: "hh'h'",
  day: "d'd' hh'h'",
  week: "d'd' hh'h'",
  month: "M'm' dd'd' hh'h'",
  year: "y'y' MM'm' dd'd' hh'h'"
}
```
Instead I was able to access it using the `_settings` property, eg: `root.durationFormatter._settings.durationFormats["second"].second`

**Comment by martynasma** [comment]
> I'm not sure if you can get the exact same result from the [amCharts 4](https://codepen.io/team/amcharts/embed/GRRoamd) example by accessing the width private setting. It would not work if you were to zoom-in or add data to the chart, because that changes the grid granularity (were labels can appear close together depending on minGridDistance property value) and not width of the chart.

Well, in that case you should use an appropriate event for those situations:
https://www.amcharts.com/docs/v5/charts/xy-chart/axes/date-axis/#events

For grid interval changes you can monitor `gridInterval` private setting:

```JavaScript
xAxis.onPrivate("gridInterval", function(gridInterval, target) {   });
```

As for the second part of your question, could you please post it into a separate issue? Let's keep threads clean with a single topic per issue.

**Comment by bye7** [comment]
> Well, in that case you should use an appropriate event for those situations

The appropriate event was axis "sizechanged" event which is fired whenever axis size changes. It is no longer available in v5.

> For grid interval changes...

You can only track timeUnit and count which is not something to base your automatic label rotations. In a dynamically updating live graph the grid distance will change without zooming-in. See "[Rotate_dynamic_update](https://github.com/amcharts/amcharts5/issues/768#issuecomment-1406026824)" video I posted. Those are the two things I would like to achieve. For now I've just set a fixed rotation.

I'm not sure whether "gridInterval" will even work for DurationAxis.

**Comment by martynasma** [comment]
> I'm not sure whether "gridInterval" will even work for DurationAxis. I'm now using a fixed rotation.

Can you try? :)

If that doesn't work, we can find another event for you.

> The appropriate event was axis "sizechanged" event which is fired whenever axis size changes. It is no longer available in v5.

Axis size does not change unless chart size changes. And that's where the `"width"` event would kick in.

If that event worked for you in v4, it was not because it was meant to, but because it was on how v4 handled axis internally.

**Comment by bye7** [comment]
> Can you try? :)

When I searched "gridInterval" on the [docs](https://www.amcharts.com/docs/v5/?s=gridInterval) I _only_ found `IDateAxisPrivate.gridInterval`, so I assumed it was only a DateAxis thing. I've just tried the following code and nothing is changed, "gridInterval" event did not fire with DurationAxis:
```JS
 var xAxis = chart.xAxes.push(
    am5xy.DurationAxis.new(root, {
      baseInterval: { timeUnit: "second", count: 15 },
      gridIntervals: [
        { timeUnit: "second", count: 15 },
        { timeUnit: "minute", count: 15 },
        { timeUnit: "day", count: 15 },
        { timeUnit: "month", count: 15 }
      ],
      renderer: am5xy.AxisRendererX.new(root, {
      }),
    })
  );
```

Even if "gridInterval" worked for DurationAxis, it is not the correct event to listen for rotating labels.

> If that doesn't work, we can find another event for you.

What I need is something like an event for [grid density](https://www.amcharts.com/docs/v5/charts/xy-chart/axes/#grid-density), so that I can rotate labels based on the "distance in pixels between any two grid lines". This will work when you zoom-in, when you add data (the grid becomes more dense) and when the width size of the chart changes, because in all of these events the grid becomes dense and we can rotate the label depending on the distance in pixels between the grid lines.

**Comment by martynasma** [comment]
Correct. `gridInterval` is for DateAxis only. I didn't realize you were still on DurationAxis, since you mentioned it does not support grouping.

How about this then:

```JavaScript
yAxis.onPrivate("step", function(step) {
  console.log(step)
});
```

**Comment by bye7** [comment]
Unfortunately the 'step' event is not correct either, it is only triggered when the interval between two consecutive tick marks on the axis change, which means you will still have labels overlapping until the grid becomes so dense and the step value increases.

**Comment by martynasma** [comment]
Isn't it what you need? Decide on whether to rotate labels based on changed number of grid lines?

**Comment by bye7** [comment]
What I want is to dynamically adjust rotation degrees, either by increasing (+1 degree) or decreasing (-1 degree) in response to continuous changes from an event (like distance in pixels between grid lines or distance in pixels between labels), as opposed to rotating the label to a fixed degree.

Watch _"Rotate_dynamic_update"_ and _"Rotate_by_width"_ from my previous [post](https://github.com/amcharts/amcharts5/issues/768#issuecomment-1406026824). Those are visual representation of what I want to achieve, watch the labels carefully. Notice one video is only adding new data and the second video is changing the chart's dimensions.

**Comment by martynasma** [comment]
Gotcha.

Your videos show resizing of the chart container. Which in turn resizes the axis. Which in turn should trigger those `width` events, you can use to determine rotation.

What am I missing here?

**Comment by bye7** [comment]
> Your videos show resizing of the chart container.

Yes, one of them does (_Rotate_by_width_). The other one does not.
I have just checked the `width` event which is <ins>only</ins> triggered when you resize the browser window, hence chart container:  `<div id="chartdiv"></div>` changes.

> What am I missing here?

You're missing two different factors that are required to take into consideration when rotating labels:
1- Zoom-in/out
2- Adding data where grid becomes denser without zooming-in and without resizing the chart container/window

**Comment by martynasma** [comment]
Gotcha.

It sounds to me that using two events (`width` and `step`) should cover all cases.

**Comment by bye7** [comment]
I believe you are not aware when exactly the events are triggered or misunderstanding the concept of dynamically rotating labels. What I want is to smoothly rotate the label based on how close they get and incrementally rotate back to its original position if there is enough space. A simple on/off toggle rotation is not what I need.

It might be a better idea if you could explain what you think should happen and the expected behaviour when using two events (`width` and `step`) you have mentioned.

Here is a demo of using them two events, they will be printed to console when they are triggered:

https://user-images.githubusercontent.com/82799248/215875675-fac7e2a3-69cb-41b5-aa74-9e61b0cd7f4b.mp4

I hope you understand the fundamental problem here.
Just in case you didn't understand the problems in the video is shown below:

Problem 1- At 0:02 The labels overlap, neither of the events have fired which means we don't know when to rotate

Problem 2- At 0:06 Let's say we arbitrarily picked a step value "4", we decided to rotate -45. When the next step is fired how do you know whether you should rotate back to 0?

Problem 3- At 0:45 After the labels are rotated to a fixed value (which is not what I want), you are unable to track whether the labels are overlapping even after you have rotated them from the step event.

**Comment by martynasma** [comment]
How about you post on CodePen what you have so far and I'll play around with it?

**Comment by martynasma** [comment]
OK, so something like this:

```JavaScript
xAxis.onPrivate("step", function(step) {
  const renderer = xAxis.get("renderer");
  const cellWidth = renderer.axisLength() * xAxis.getCellWidthPosition() * am5.time.getIntervalDuration(xAxis.getPrivate("gridInterval")) / xAxis.baseDuration();
  // calculate rotation based on cell width
  const rotation = 90; // ....
  xAxis.get("renderer").labels.template.set("rotation", rotation);
})
```

**Comment by bye7** [comment]
> What I have is <ins>almost</ins> identical to "[Line Chart Adding Data Every Second Demo](https://www.amcharts.com/demos/line-chart-adding-data-every-second/)" <ins>except</ins> I don't delete the first array element of the series.data and <ins>I use DurationAxis instead of DateAxis</ins>.

* The "step" event does not work the same on "**DurationAxis**" as it does in "**DateAxis**".
* The code below does not work in DurationAxis, because `baseDuration()` method does not work for DurationAxis.

Here is what I have **<ins>for DateAxis</ins>**, just to demonstrate what I would like to achieve (even though there are some issues with it):

```JS
am5.ready(function () {
    xAxis.onPrivate("step", function (step) {
        const xRenderer = xAxis.get("renderer");
        const gridInterval = xAxis.getPrivate("gridInterval");
        const cellWidth = xRenderer.axisLength() * xAxis.getCellWidthPosition() * am5.time.getIntervalDuration(gridInterval) / xAxis.baseDuration();
        
        /*
        There has to be a better way to get the maximum label width among the displayed x-axis labels without setting
        maxWidth to a static value in xRenderer.labels.template. Because, each time the ".timeUnit" changes from second
        to minute, it has a different format that will cause the label width to increase or decrease.
        Example: "ss" to "mm:ss", the label width increased.
        */
        let labelWidth;
        switch (gridInterval.timeUnit) {
            case "millisecond":
                // set labelWidth
                break;
            case "second":
                labelWidth = 65; // Based on the default label date formats "HH:mm:ss" we assume the pixel width will be 65
                break;
            case "minute":
                labelWidth = 55; // Based on the default label date formats "HH:mm" we assume the pixel width will be 55
                break;
            case "hour":
                labelWidth = 55; // Based on the default label date formats "HH:mm" we assume the pixel width will be 55
                break;
            case "day":
                // set labelWidth
                break;
            case "week":
                // set labelWidth
                break;
            case "month":
                // set labelWidth
                break;
            case "year":
                // set labelWidth
                break;
            default:
                labelWidth = 65;
        }
        if (cellWidth < labelWidth) {
            const angle = cellWidth - labelWidth;
            if (angle >= -90 && angle <= 0) {
                xRenderer.labels.template.set("rotation", angle);
            }

        }
    });
})
```


https://user-images.githubusercontent.com/82799248/216209148-d3876127-08cf-46b6-9096-01c017602738.mp4

See how much better this is compared to just overlapping the text?
This should be the default behaviour like how it is on [Chart.JS](https://github.com/chartjs/Chart.js/blob/1684cf7dbf577d97fe20f6039e895388ed15902c/src/core/core.scale.js#L585-L622)

How can I do the same for DurationAxis ?

**Comment by martynasma** [comment]
We're adding `cellWidth` private setting to all axes. You will be able to use its events and value for your needs.

I'll let you know when the new version is out.

**Comment by martynasma** [comment]
Fixed in [5.3.4](https://github.com/amcharts/amcharts5/releases/tag/5.3.4).

## [5.3.4] - 2023-02-20

### Added
- `cellWidth` read-only private setting added to `Axis`. Holds a distance in pixels between two grid lines. It might not be accurate as grid on `DateAxis` might not be at equal intervals.
- `getCellWidthPosition()` added to all axis types. Returns relative distance between two grid lines of the axis.

### Fixed
- HTML-based tooltip might produce a funky animation if it was displayed very close to the edge of the chart.
- `StockChart`'s X-axes from all panels not always were syncing properly between data updates.
- Fixing bug where click event was not dispatched on tooltips outside of the chart boundaries.
- Changing a color for a `LineSeries` in with a `StockChart` control was not changing its tooltip color.
- `GaplessDateAxis` was not taking into account missing dates when using `zoomToDates()` or `zoomToValues()` methods.
- When data grouping was on and data was changed, `ValueAxis` was zooming to incorrect position.
- Radial labels sometimes could get reversed when updating their settings.

[Full change log](https://github.com/amcharts/amcharts5/blob/master/packages/shared/CHANGELOG.md).

[Download options](https://www.amcharts.com/download/).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.

**Comment by martynasma** [comment]
OK, so now you can do this:

```JavaScript
xAxis.onPrivate("cellWidth", function (width) {
  // ...
});
```

It will work on any axis type.
